# ceQT: A co-expression QTL model to reveal mediators between variants and target genes

## Abstract

The expression Quantitative Trait Locus (eQTL) mapping is used to infer the relationship between genotypes and gene expressions. Nowadays, it has become a routine analysis for projects with both genotype and expression data available to find the associations between variant genotypes and gene expression values, and a quite number of databases have been built based on up it. Those databases have collected or computed numerous of the variant-gene pairs, of which the gene's expression values are significantly associated with the genotypes of the variant. However, these associations do not give any clues on how and why they exist. To address this issue, here we present a new QTL mapping model, co-expression QTL (ceQTL), which investigates how the variant genotype is affecting the co-expression between a mediator (i.e. a transcription factor, TF) and the target gene. Since ceQTL requires no more types of data than eQTL does,  it can be applied to any datasets where eQTL is applicable. One of the concerns of this model might be the search space, thus the huge number of possible mediator-gene pairs to consider. To overcome that, we only take into account the cis-ceQTLs, landing in Â±250,000 bp flanking regions of the target genes, and limit the TF-gene pairs by a significant computational binding affinity of the TF to such region of a gene using TF binding motif scanning. In this manuscript, the ceQTL model is applied to the LUAD dataset from TCGA. We found that ceQTL can recover most of the variant-gene pairs that are discovered by the traditional eQTL model. Mediation analysis also shows significant median effects for the TF to mediate the relationship between the variant and the target gene. We also observed that the TF binding affinity on the target gene is significantly affected by the genotype of the variant. All these observations show that ceQTL is an enhanced eQTL model, which enables the possibility for us to get additional information on how genotypes and expressions are associated, hence provides more opportunities for us to understand the mechanisms of diseases, where these genotype-expression relationships have been observed.

